bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerMetadataSerDe.java

```diff
@@ -51,122 +58,267 @@
+                break;
+            } else {
+                versionBuf[i++] = (byte) b;
+            }
         }
-
-        return s.toString().getBytes(UTF_8);
+        throw new IOException("Unable to find end of version number, metadata appears corrupt");
     }
 
-    @VisibleForTesting
-    public LedgerMetadataFormat buildProtoFormat(LedgerMetadata metadata) {
-        LedgerMetadataFormat.Builder builder = LedgerMetadataFormat.newBuilder();
-        builder.setQuorumSize(metadata.getWriteQuorumSize())
-            .setAckQuorumSize(metadata.getAckQuorumSize())
-            .setEnsembleSize(metadata.getEnsembleSize())
-            .setLength(metadata.getLength())
-            .setLastEntryId(metadata.getLastEntryId());
-
-        switch (metadata.getState()) {
-        case CLOSED:
-            builder.setState(LedgerMetadataFormat.State.CLOSED);
+    public byte[] serialize(LedgerMetadata metadata) throws IOException {
+        int formatVersion = metadata.getMetadataFormatVersion();
+        final byte[] serialized;
+        switch (formatVersion) {
+        case METADATA_FORMAT_VERSION_3:
+            serialized = serializeVersion3(metadata);
             break;
-        case IN_RECOVERY:
-            builder.setState(LedgerMetadataFormat.State.IN_RECOVERY);
+        case METADATA_FORMAT_VERSION_2:
+            serialized = serializeVersion2(metadata);
             break;
-        case OPEN:
-            builder.setState(LedgerMetadataFormat.State.OPEN);
+        case METADATA_FORMAT_VERSION_1:
+            serialized = serializeVersion1(metadata);
             break;
         default:
-            checkArgument(false,
-                          String.format("Unknown state %s for protobuf serialization", metadata.getState()));
-            break;
+            throw new IllegalArgumentException("Invalid format version " + formatVersion);
+        }
+        if (log.isDebugEnabled()) {
+            String serializedStr;
+            if (formatVersion > METADATA_FORMAT_VERSION_2) {
+                serializedStr = Base64.getEncoder().encodeToString(serialized);
+            } else {
+                serializedStr = new String(serialized, UTF_8);
+            }
+            log.debug("Serialized with format {}: {}", formatVersion, serializedStr);
         }
+        return serialized;
+    }
+
+    private static byte[] serializeVersion3(LedgerMetadata metadata) throws IOException {
+        try (ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+            writeHeader(os, METADATA_FORMAT_VERSION_3);
+            LedgerMetadataFormat.Builder builder = LedgerMetadataFormat.newBuilder();
+            builder.setQuorumSize(metadata.getWriteQuorumSize())
+                .setAckQuorumSize(metadata.getAckQuorumSize())
+                .setEnsembleSize(metadata.getEnsembleSize())
+                .setLength(metadata.getLength())
+                .setLastEntryId(metadata.getLastEntryId());
+
+            switch (metadata.getState()) {
```

Comment:
All these duplicates are on purpose. For example, if we add a state in future, but the max metadata version is 2, and we are using shared code, it will happily write the new state to the metadata and break clients that can only read V2. It's ugly, but it's here defensively.

Reply:
For ‘state’, it is probably the case. However I am not sure about the other comments.

Reply:
A field could be added to LedgerMetadataFormat.Segment or LedgerMetadataFormat.cMetadataMapEntry, which could cause the same problem. Will factor out password though.

Reply:
ack